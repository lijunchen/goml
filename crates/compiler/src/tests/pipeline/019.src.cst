FILE@0..1264
  STRUCT@0..42
    StructKeyword@0..6 "struct"
    Whitespace@6..7 " "
    Uident@7..12 "Point"
    Whitespace@12..13 " "
    LBrace@13..14 "{"
    Whitespace@14..19 "\n    "
    STRUCT_FIELD_LIST@19..42
      STRUCT_FIELD@19..25
        Lident@19..20 "x"
        Colon@20..21 ":"
        Whitespace@21..22 " "
        TYPE_INT@22..25
          IntKeyword@22..25 "int"
      Comma@25..26 ","
      Whitespace@26..31 "\n    "
      STRUCT_FIELD@31..37
        Lident@31..32 "y"
        Colon@32..33 ":"
        Whitespace@33..34 " "
        TYPE_INT@34..37
          IntKeyword@34..37 "int"
      Comma@37..38 ","
      Whitespace@38..39 "\n"
      RBrace@39..40 "}"
      Whitespace@40..42 "\n\n"
  STRUCT@42..79
    StructKeyword@42..48 "struct"
    Whitespace@48..49 " "
    Uident@49..56 "Wrapper"
    GENERIC_LIST@56..60
      LBracket@56..57 "["
      GENERIC@57..58
        Uident@57..58 "T"
      RBracket@58..59 "]"
      Whitespace@59..60 " "
    LBrace@60..61 "{"
    Whitespace@61..66 "\n    "
    STRUCT_FIELD_LIST@66..79
      STRUCT_FIELD@66..74
        Lident@66..71 "value"
        Colon@71..72 ":"
        Whitespace@72..73 " "
        TYPE_TAPP@73..74
          Uident@73..74 "T"
      Comma@74..75 ","
      Whitespace@75..76 "\n"
      RBrace@76..77 "}"
      Whitespace@77..79 "\n\n"
  FN@79..134
    FnKeyword@79..81 "fn"
    Whitespace@81..82 " "
    Lident@82..92 "make_point"
    PARAM_LIST@92..95
      LParen@92..93 "("
      RParen@93..94 ")"
      Whitespace@94..95 " "
    Arrow@95..97 "->"
    Whitespace@97..98 " "
    TYPE_TAPP@98..104
      Uident@98..103 "Point"
      Whitespace@103..104 " "
    BLOCK@104..134
      LBrace@104..105 "{"
      Whitespace@105..110 "\n    "
      EXPR_STRUCT_LITERAL@110..131
        Uident@110..115 "Point"
        Whitespace@115..116 " "
        STRUCT_LITERAL_FIELD_LIST@116..131
          LBrace@116..117 "{"
          Whitespace@117..118 " "
          STRUCT_LITERAL_FIELD@118..122
            Lident@118..119 "x"
            Colon@119..120 ":"
            Whitespace@120..121 " "
            EXPR_INT@121..122
              Int@121..122 "0"
          Comma@122..123 ","
          Whitespace@123..124 " "
          STRUCT_LITERAL_FIELD@124..129
            Lident@124..125 "y"
            Colon@125..126 ":"
            Whitespace@126..127 " "
            EXPR_INT@127..129
              Int@127..128 "0"
              Whitespace@128..129 " "
          RBrace@129..130 "}"
          Whitespace@130..131 "\n"
      RBrace@131..132 "}"
      Whitespace@132..134 "\n\n"
  FN@134..235
    FnKeyword@134..136 "fn"
    Whitespace@136..137 " "
    Lident@137..141 "flip"
    PARAM_LIST@141..156
      LParen@141..142 "("
      PARAM@142..154
        Lident@142..147 "point"
        Colon@147..148 ":"
        Whitespace@148..149 " "
        TYPE_TAPP@149..154
          Uident@149..154 "Point"
      RParen@154..155 ")"
      Whitespace@155..156 " "
    Arrow@156..158 "->"
    Whitespace@158..159 " "
    TYPE_TAPP@159..165
      Uident@159..164 "Point"
      Whitespace@164..165 " "
    BLOCK@165..235
      LBrace@165..166 "{"
      Whitespace@166..171 "\n    "
      EXPR_LET@171..232
        LetKeyword@171..174 "let"
        Whitespace@174..175 " "
        PATTERN_CONSTR@175..196
          Uident@175..180 "Point"
          Whitespace@180..181 " "
          STRUCT_PATTERN_FIELD_LIST@181..196
            LBrace@181..182 "{"
            Whitespace@182..183 " "
            STRUCT_PATTERN_FIELD@183..187
              Lident@183..184 "x"
              Colon@184..185 ":"
              Whitespace@185..186 " "
              PATTERN_VARIABLE@186..187
                Lident@186..187 "x"
            Comma@187..188 ","
            Whitespace@188..189 " "
            STRUCT_PATTERN_FIELD@189..194
              Lident@189..190 "y"
              Colon@190..191 ":"
              Whitespace@191..192 " "
              PATTERN_VARIABLE@192..194
                Lident@192..193 "y"
                Whitespace@193..194 " "
            RBrace@194..195 "}"
            Whitespace@195..196 " "
        Eq@196..197 "="
        Whitespace@197..198 " "
        EXPR_LET_VALUE@198..204
          EXPR_LIDENT@198..204
            Lident@198..203 "point"
            Whitespace@203..204 " "
        InKeyword@204..206 "in"
        Whitespace@206..211 "\n    "
        EXPR_LET_BODY@211..232
          EXPR_STRUCT_LITERAL@211..232
            Uident@211..216 "Point"
            Whitespace@216..217 " "
            STRUCT_LITERAL_FIELD_LIST@217..232
              LBrace@217..218 "{"
              Whitespace@218..219 " "
              STRUCT_LITERAL_FIELD@219..223
                Lident@219..220 "y"
                Colon@220..221 ":"
                Whitespace@221..222 " "
                EXPR_LIDENT@222..223
                  Lident@222..223 "x"
              Comma@223..224 ","
              Whitespace@224..225 " "
              STRUCT_LITERAL_FIELD@225..230
                Lident@225..226 "x"
                Colon@226..227 ":"
                Whitespace@227..228 " "
                EXPR_LIDENT@228..230
                  Lident@228..229 "y"
                  Whitespace@229..230 " "
              RBrace@230..231 "}"
              Whitespace@231..232 "\n"
      RBrace@232..233 "}"
      Whitespace@233..235 "\n\n"
  FN@235..301
    FnKeyword@235..237 "fn"
    Whitespace@237..238 " "
    Lident@238..246 "wrap_int"
    PARAM_LIST@246..255
      LParen@246..247 "("
      PARAM@247..253
        Lident@247..248 "x"
        Colon@248..249 ":"
        Whitespace@249..250 " "
        TYPE_INT@250..253
          IntKeyword@250..253 "int"
      RParen@253..254 ")"
      Whitespace@254..255 " "
    Arrow@255..257 "->"
    Whitespace@257..258 " "
    TYPE_TAPP@258..271
      Uident@258..265 "Wrapper"
      TYPE_PARAM_LIST@265..271
        LBracket@265..266 "["
        TYPE_INT@266..269
          IntKeyword@266..269 "int"
        RBracket@269..270 "]"
        Whitespace@270..271 " "
    BLOCK@271..301
      LBrace@271..272 "{"
      Whitespace@272..277 "\n    "
      EXPR_STRUCT_LITERAL@277..298
        Uident@277..284 "Wrapper"
        Whitespace@284..285 " "
        STRUCT_LITERAL_FIELD_LIST@285..298
          LBrace@285..286 "{"
          Whitespace@286..287 " "
          STRUCT_LITERAL_FIELD@287..296
            Lident@287..292 "value"
            Colon@292..293 ":"
            Whitespace@293..294 " "
            EXPR_LIDENT@294..296
              Lident@294..295 "x"
              Whitespace@295..296 " "
          RBrace@296..297 "}"
          Whitespace@297..298 "\n"
      RBrace@298..299 "}"
      Whitespace@299..301 "\n\n"
  FN@301..401
    FnKeyword@301..303 "fn"
    Whitespace@303..304 " "
    Lident@304..311 "x_add_1"
    PARAM_LIST@311..322
      LParen@311..312 "("
      PARAM@312..320
        Lident@312..313 "p"
        Colon@313..314 ":"
        Whitespace@314..315 " "
        TYPE_TAPP@315..320
          Uident@315..320 "Point"
      RParen@320..321 ")"
      Whitespace@321..322 " "
    Arrow@322..324 "->"
    Whitespace@324..325 " "
    TYPE_TAPP@325..331
      Uident@325..330 "Point"
      Whitespace@330..331 " "
    BLOCK@331..401
      LBrace@331..332 "{"
      Whitespace@332..337 "\n    "
      EXPR_LET@337..398
        LetKeyword@337..340 "let"
        Whitespace@340..341 " "
        PATTERN_CONSTR@341..362
          Uident@341..346 "Point"
          Whitespace@346..347 " "
          STRUCT_PATTERN_FIELD_LIST@347..362
            LBrace@347..348 "{"
            Whitespace@348..349 " "
            STRUCT_PATTERN_FIELD@349..353
              Lident@349..350 "x"
              Colon@350..351 ":"
              Whitespace@351..352 " "
              PATTERN_VARIABLE@352..353
                Lident@352..353 "x"
            Comma@353..354 ","
            Whitespace@354..355 " "
            STRUCT_PATTERN_FIELD@355..360
              Lident@355..356 "y"
              Colon@356..357 ":"
              Whitespace@357..358 " "
              PATTERN_VARIABLE@358..360
                Lident@358..359 "y"
                Whitespace@359..360 " "
            RBrace@360..361 "}"
            Whitespace@361..362 " "
        Eq@362..363 "="
        Whitespace@363..364 " "
        EXPR_LET_VALUE@364..366
          EXPR_LIDENT@364..366
            Lident@364..365 "p"
            Whitespace@365..366 " "
        InKeyword@366..368 "in"
        Whitespace@368..373 "\n    "
        EXPR_LET_BODY@373..398
          EXPR_STRUCT_LITERAL@373..398
            Uident@373..378 "Point"
            Whitespace@378..379 " "
            STRUCT_LITERAL_FIELD_LIST@379..398
              LBrace@379..380 "{"
              Whitespace@380..381 " "
              STRUCT_LITERAL_FIELD@381..389
                Lident@381..382 "x"
                Colon@382..383 ":"
                Whitespace@383..384 " "
                EXPR_BINARY@384..389
                  EXPR_LIDENT@384..386
                    Lident@384..385 "x"
                    Whitespace@385..386 " "
                  Plus@386..387 "+"
                  Whitespace@387..388 " "
                  EXPR_INT@388..389
                    Int@388..389 "1"
              Comma@389..390 ","
              Whitespace@390..391 " "
              STRUCT_LITERAL_FIELD@391..396
                Lident@391..392 "y"
                Colon@392..393 ":"
                Whitespace@393..394 " "
                EXPR_LIDENT@394..396
                  Lident@394..395 "y"
                  Whitespace@395..396 " "
              RBrace@396..397 "}"
              Whitespace@397..398 "\n"
      RBrace@398..399 "}"
      Whitespace@399..401 "\n\n"
  FN@401..474
    FnKeyword@401..403 "fn"
    Whitespace@403..404 " "
    Lident@404..409 "get_x"
    PARAM_LIST@409..420
      LParen@409..410 "("
      PARAM@410..418
        Lident@410..411 "p"
        Colon@411..412 ":"
        Whitespace@412..413 " "
        TYPE_TAPP@413..418
          Uident@413..418 "Point"
      RParen@418..419 ")"
      Whitespace@419..420 " "
    Arrow@420..422 "->"
    Whitespace@422..423 " "
    TYPE_INT@423..427
      IntKeyword@423..426 "int"
      Whitespace@426..427 " "
    BLOCK@427..474
      LBrace@427..428 "{"
      Whitespace@428..433 "\n    "
      EXPR_LET@433..471
        LetKeyword@433..436 "let"
        Whitespace@436..437 " "
        PATTERN_CONSTR@437..458
          Uident@437..442 "Point"
          Whitespace@442..443 " "
          STRUCT_PATTERN_FIELD_LIST@443..458
            LBrace@443..444 "{"
            Whitespace@444..445 " "
            STRUCT_PATTERN_FIELD@445..449
              Lident@445..446 "x"
              Colon@446..447 ":"
              Whitespace@447..448 " "
              PATTERN_VARIABLE@448..449
                Lident@448..449 "x"
            Comma@449..450 ","
            Whitespace@450..451 " "
            STRUCT_PATTERN_FIELD@451..456
              Lident@451..452 "y"
              Colon@452..453 ":"
              Whitespace@453..454 " "
              PATTERN_VARIABLE@454..456
                Lident@454..455 "y"
                Whitespace@455..456 " "
            RBrace@456..457 "}"
            Whitespace@457..458 " "
        Eq@458..459 "="
        Whitespace@459..460 " "
        EXPR_LET_VALUE@460..462
          EXPR_LIDENT@460..462
            Lident@460..461 "p"
            Whitespace@461..462 " "
        InKeyword@462..464 "in"
        Whitespace@464..469 "\n    "
        EXPR_LET_BODY@469..471
          EXPR_LIDENT@469..471
            Lident@469..470 "x"
            Whitespace@470..471 "\n"
      RBrace@471..472 "}"
      Whitespace@472..474 "\n\n"
  FN@474..547
    FnKeyword@474..476 "fn"
    Whitespace@476..477 " "
    Lident@477..482 "get_y"
    PARAM_LIST@482..493
      LParen@482..483 "("
      PARAM@483..491
        Lident@483..484 "p"
        Colon@484..485 ":"
        Whitespace@485..486 " "
        TYPE_TAPP@486..491
          Uident@486..491 "Point"
      RParen@491..492 ")"
      Whitespace@492..493 " "
    Arrow@493..495 "->"
    Whitespace@495..496 " "
    TYPE_INT@496..500
      IntKeyword@496..499 "int"
      Whitespace@499..500 " "
    BLOCK@500..547
      LBrace@500..501 "{"
      Whitespace@501..506 "\n    "
      EXPR_LET@506..544
        LetKeyword@506..509 "let"
        Whitespace@509..510 " "
        PATTERN_CONSTR@510..531
          Uident@510..515 "Point"
          Whitespace@515..516 " "
          STRUCT_PATTERN_FIELD_LIST@516..531
            LBrace@516..517 "{"
            Whitespace@517..518 " "
            STRUCT_PATTERN_FIELD@518..522
              Lident@518..519 "x"
              Colon@519..520 ":"
              Whitespace@520..521 " "
              PATTERN_VARIABLE@521..522
                Lident@521..522 "x"
            Comma@522..523 ","
            Whitespace@523..524 " "
            STRUCT_PATTERN_FIELD@524..529
              Lident@524..525 "y"
              Colon@525..526 ":"
              Whitespace@526..527 " "
              PATTERN_VARIABLE@527..529
                Lident@527..528 "y"
                Whitespace@528..529 " "
            RBrace@529..530 "}"
            Whitespace@530..531 " "
        Eq@531..532 "="
        Whitespace@532..533 " "
        EXPR_LET_VALUE@533..535
          EXPR_LIDENT@533..535
            Lident@533..534 "p"
            Whitespace@534..535 " "
        InKeyword@535..537 "in"
        Whitespace@537..542 "\n    "
        EXPR_LET_BODY@542..544
          EXPR_LIDENT@542..544
            Lident@542..543 "y"
            Whitespace@543..544 "\n"
      RBrace@544..545 "}"
      Whitespace@545..547 "\n\n"
  FN@547..801
    FnKeyword@547..549 "fn"
    Whitespace@549..550 " "
    Lident@550..565 "point_to_string"
    PARAM_LIST@565..576
      LParen@565..566 "("
      PARAM@566..574
        Lident@566..567 "p"
        Colon@567..568 ":"
        Whitespace@568..569 " "
        TYPE_TAPP@569..574
          Uident@569..574 "Point"
      RParen@574..575 ")"
      Whitespace@575..576 " "
    Arrow@576..578 "->"
    Whitespace@578..579 " "
    TYPE_STRING@579..586
      StringKeyword@579..585 "string"
      Whitespace@585..586 " "
    BLOCK@586..801
      LBrace@586..587 "{"
      Whitespace@587..592 "\n    "
      EXPR_LET@592..798
        LetKeyword@592..595 "let"
        Whitespace@595..596 " "
        PATTERN_CONSTR@596..617
          Uident@596..601 "Point"
          Whitespace@601..602 " "
          STRUCT_PATTERN_FIELD_LIST@602..617
            LBrace@602..603 "{"
            Whitespace@603..604 " "
            STRUCT_PATTERN_FIELD@604..608
              Lident@604..605 "x"
              Colon@605..606 ":"
              Whitespace@606..607 " "
              PATTERN_VARIABLE@607..608
                Lident@607..608 "x"
            Comma@608..609 ","
            Whitespace@609..610 " "
            STRUCT_PATTERN_FIELD@610..615
              Lident@610..611 "y"
              Colon@611..612 ":"
              Whitespace@612..613 " "
              PATTERN_VARIABLE@613..615
                Lident@613..614 "y"
                Whitespace@614..615 " "
            RBrace@615..616 "}"
            Whitespace@616..617 " "
        Eq@617..618 "="
        Whitespace@618..619 " "
        EXPR_LET_VALUE@619..621
          EXPR_LIDENT@619..621
            Lident@619..620 "p"
            Whitespace@620..621 " "
        InKeyword@621..623 "in"
        Whitespace@623..628 "\n    "
        EXPR_LET_BODY@628..798
          EXPR_CALL@628..798
            EXPR_LIDENT@628..638
              Lident@628..638 "string_add"
            ARG_LIST@638..798
              LParen@638..639 "("
              Whitespace@639..648 "\n        "
              ARG@648..787
                EXPR_CALL@648..777
                  EXPR_LIDENT@648..658
                    Lident@648..658 "string_add"
                  ARG_LIST@658..777
                    LParen@658..659 "("
                    Whitespace@659..672 "\n            "
                    ARG@672..729
                      EXPR_CALL@672..715
                        EXPR_LIDENT@672..682
                          Lident@672..682 "string_add"
                        ARG_LIST@682..715
                          LParen@682..683 "("
                          ARG@683..698
                            EXPR_STR@683..696
                              Str@683..696 "\"Point { x: \""
                            Comma@696..697 ","
                            Whitespace@697..698 " "
                          ARG@698..714
                            EXPR_CALL@698..714
                              EXPR_LIDENT@698..711
                                Lident@698..711 "int_to_string"
                              ARG_LIST@711..714
                                LParen@711..712 "("
                                ARG@712..713
                                  EXPR_LIDENT@712..713
                                    Lident@712..713 "x"
                                RParen@713..714 ")"
                          RParen@714..715 ")"
                      Comma@715..716 ","
                      Whitespace@716..729 "\n            "
                    ARG@729..776
                      EXPR_CALL@729..766
                        EXPR_LIDENT@729..739
                          Lident@729..739 "string_add"
                        ARG_LIST@739..766
                          LParen@739..740 "("
                          ARG@740..749
                            EXPR_STR@740..747
                              Str@740..747 "\", y: \""
                            Comma@747..748 ","
                            Whitespace@748..749 " "
                          ARG@749..765
                            EXPR_CALL@749..765
                              EXPR_LIDENT@749..762
                                Lident@749..762 "int_to_string"
                              ARG_LIST@762..765
                                LParen@762..763 "("
                                ARG@763..764
                                  EXPR_LIDENT@763..764
                                    Lident@763..764 "y"
                                RParen@764..765 ")"
                          RParen@765..766 ")"
                      Comma@766..767 ","
                      Whitespace@767..776 "\n        "
                    RParen@776..777 ")"
                Comma@777..778 ","
                Whitespace@778..787 "\n        "
              ARG@787..796
                EXPR_STR@787..790
                  Str@787..790 "\"}\""
                Comma@790..791 ","
                Whitespace@791..796 "\n    "
              RParen@796..797 ")"
              Whitespace@797..798 "\n"
      RBrace@798..799 "}"
      Whitespace@799..801 "\n\n"
  FN@801..1264
    FnKeyword@801..803 "fn"
    Whitespace@803..804 " "
    Lident@804..808 "main"
    PARAM_LIST@808..811
      LParen@808..809 "("
      RParen@809..810 ")"
      Whitespace@810..811 " "
    BLOCK@811..1264
      LBrace@811..812 "{"
      Whitespace@812..817 "\n    "
      EXPR_LET@817..1262
        LetKeyword@817..820 "let"
        Whitespace@820..821 " "
        PATTERN_VARIABLE@821..827
          Lident@821..826 "start"
          Whitespace@826..827 " "
        Eq@827..828 "="
        Whitespace@828..829 " "
        EXPR_LET_VALUE@829..842
          EXPR_CALL@829..842
            EXPR_LIDENT@829..839
              Lident@829..839 "make_point"
            ARG_LIST@839..842
              LParen@839..840 "("
              RParen@840..841 ")"
              Whitespace@841..842 " "
        InKeyword@842..844 "in"
        Whitespace@844..849 "\n    "
        EXPR_LET_BODY@849..1262
          EXPR_LET@849..1262
            LetKeyword@849..852 "let"
            Whitespace@852..853 " "
            PATTERN_WILDCARD@853..855
              WildcardKeyword@853..854 "_"
              Whitespace@854..855 " "
            Eq@855..856 "="
            Whitespace@856..857 " "
            EXPR_LET_VALUE@857..896
              EXPR_CALL@857..896
                EXPR_LIDENT@857..871
                  Lident@857..871 "string_println"
                ARG_LIST@871..896
                  LParen@871..872 "("
                  ARG@872..894
                    EXPR_CALL@872..894
                      EXPR_LIDENT@872..887
                        Lident@872..887 "point_to_string"
                      ARG_LIST@887..894
                        LParen@887..888 "("
                        ARG@888..893
                          EXPR_LIDENT@888..893
                            Lident@888..893 "start"
                        RParen@893..894 ")"
                  RParen@894..895 ")"
                  Whitespace@895..896 " "
            InKeyword@896..898 "in"
            Whitespace@898..904 "\n\n    "
            EXPR_LET_BODY@904..1262
              EXPR_LET@904..1262
                LetKeyword@904..907 "let"
                Whitespace@907..908 " "
                PATTERN_VARIABLE@908..916
                  Lident@908..915 "swapped"
                  Whitespace@915..916 " "
                Eq@916..917 "="
                Whitespace@917..918 " "
                EXPR_LET_VALUE@918..945
                  EXPR_CALL@918..945
                    EXPR_LIDENT@918..922
                      Lident@918..922 "flip"
                    ARG_LIST@922..945
                      LParen@922..923 "("
                      ARG@923..943
                        EXPR_STRUCT_LITERAL@923..943
                          Uident@923..928 "Point"
                          Whitespace@928..929 " "
                          STRUCT_LITERAL_FIELD_LIST@929..943
                            LBrace@929..930 "{"
                            Whitespace@930..931 " "
                            STRUCT_LITERAL_FIELD@931..935
                              Lident@931..932 "y"
                              Colon@932..933 ":"
                              Whitespace@933..934 " "
                              EXPR_INT@934..935
                                Int@934..935 "2"
                            Comma@935..936 ","
                            Whitespace@936..937 " "
                            STRUCT_LITERAL_FIELD@937..942
                              Lident@937..938 "x"
                              Colon@938..939 ":"
                              Whitespace@939..940 " "
                              EXPR_INT@940..942
                                Int@940..941 "1"
                                Whitespace@941..942 " "
                            RBrace@942..943 "}"
                      RParen@943..944 ")"
                      Whitespace@944..945 " "
                InKeyword@945..947 "in"
                Whitespace@947..952 "\n    "
                EXPR_LET_BODY@952..1262
                  EXPR_LET@952..1262
                    LetKeyword@952..955 "let"
                    Whitespace@955..956 " "
                    PATTERN_WILDCARD@956..958
                      WildcardKeyword@956..957 "_"
                      Whitespace@957..958 " "
                    Eq@958..959 "="
                    Whitespace@959..960 " "
                    EXPR_LET_VALUE@960..1001
                      EXPR_CALL@960..1001
                        EXPR_LIDENT@960..974
                          Lident@960..974 "string_println"
                        ARG_LIST@974..1001
                          LParen@974..975 "("
                          ARG@975..999
                            EXPR_CALL@975..999
                              EXPR_LIDENT@975..990
                                Lident@975..990 "point_to_string"
                              ARG_LIST@990..999
                                LParen@990..991 "("
                                ARG@991..998
                                  EXPR_LIDENT@991..998
                                    Lident@991..998 "swapped"
                                RParen@998..999 ")"
                          RParen@999..1000 ")"
                          Whitespace@1000..1001 " "
                    InKeyword@1001..1003 "in"
                    Whitespace@1003..1009 "\n\n    "
                    EXPR_LET_BODY@1009..1262
                      EXPR_LET@1009..1262
                        LetKeyword@1009..1012 "let"
                        Whitespace@1012..1013 " "
                        PATTERN_VARIABLE@1013..1019
                          Lident@1013..1018 "boxed"
                          Whitespace@1018..1019 " "
                        Eq@1019..1020 "="
                        Whitespace@1020..1021 " "
                        EXPR_LET_VALUE@1021..1033
                          EXPR_CALL@1021..1033
                            EXPR_LIDENT@1021..1029
                              Lident@1021..1029 "wrap_int"
                            ARG_LIST@1029..1033
                              LParen@1029..1030 "("
                              ARG@1030..1031
                                EXPR_INT@1030..1031
                                  Int@1030..1031 "3"
                              RParen@1031..1032 ")"
                              Whitespace@1032..1033 " "
                        InKeyword@1033..1035 "in"
                        Whitespace@1035..1040 "\n    "
                        EXPR_LET_BODY@1040..1262
                          EXPR_LET@1040..1262
                            LetKeyword@1040..1043 "let"
                            Whitespace@1043..1044 " "
                            PATTERN_VARIABLE@1044..1054
                              Lident@1044..1053 "point_box"
                              Whitespace@1053..1054 " "
                            Eq@1054..1055 "="
                            Whitespace@1055..1056 " "
                            EXPR_LET_VALUE@1056..1083
                              EXPR_STRUCT_LITERAL@1056..1083
                                Uident@1056..1063 "Wrapper"
                                Whitespace@1063..1064 " "
                                STRUCT_LITERAL_FIELD_LIST@1064..1083
                                  LBrace@1064..1065 "{"
                                  Whitespace@1065..1066 " "
                                  STRUCT_LITERAL_FIELD@1066..1081
                                    Lident@1066..1071 "value"
                                    Colon@1071..1072 ":"
                                    Whitespace@1072..1073 " "
                                    EXPR_LIDENT@1073..1081
                                      Lident@1073..1080 "swapped"
                                      Whitespace@1080..1081 " "
                                  RBrace@1081..1082 "}"
                                  Whitespace@1082..1083 " "
                            InKeyword@1083..1085 "in"
                            Whitespace@1085..1091 "\n\n    "
                            EXPR_LET_BODY@1091..1262
                              EXPR_LET@1091..1262
                                LetKeyword@1091..1094 "let"
                                Whitespace@1094..1095 " "
                                PATTERN_VARIABLE@1095..1097
                                  Lident@1095..1096 "a"
                                  Whitespace@1096..1097 " "
                                Eq@1097..1098 "="
                                Whitespace@1098..1099 " "
                                EXPR_LET_VALUE@1099..1114
                                  EXPR_CALL@1099..1114
                                    EXPR_LIDENT@1099..1106
                                      Lident@1099..1106 "x_add_1"
                                    ARG_LIST@1106..1114
                                      LParen@1106..1107 "("
                                      ARG@1107..1112
                                        EXPR_LIDENT@1107..1112
                                          Lident@1107..1112 "start"
                                      RParen@1112..1113 ")"
                                      Whitespace@1113..1114 " "
                                InKeyword@1114..1116 "in"
                                Whitespace@1116..1121 "\n    "
                                EXPR_LET_BODY@1121..1262
                                  EXPR_LET@1121..1262
                                    LetKeyword@1121..1124 "let"
                                    Whitespace@1124..1125 " "
                                    PATTERN_WILDCARD@1125..1127
                                      WildcardKeyword@1125..1126 "_"
                                      Whitespace@1126..1127 " "
                                    Eq@1127..1128 "="
                                    Whitespace@1128..1129 " "
                                    EXPR_LET_VALUE@1129..1164
                                      EXPR_CALL@1129..1164
                                        EXPR_LIDENT@1129..1143
                                          Lident@1129..1143 "string_println"
                                        ARG_LIST@1143..1164
                                          LParen@1143..1144 "("
                                          ARG@1144..1162
                                            EXPR_CALL@1144..1162
                                              EXPR_LIDENT@1144..1159
                                                Lident@1144..1159 "point_to_string"
                                              ARG_LIST@1159..1162
                                                LParen@1159..1160 "("
                                                ARG@1160..1161
                                                  EXPR_LIDENT@1160..1161
                                                    Lident@1160..1161 "a"
                                                RParen@1161..1162 ")"
                                          RParen@1162..1163 ")"
                                          Whitespace@1163..1164 " "
                                    InKeyword@1164..1166 "in"
                                    Whitespace@1166..1172 "\n\n    "
                                    EXPR_LET_BODY@1172..1262
                                      EXPR_LET@1172..1262
                                        LetKeyword@1172..1175 "let"
                                        Whitespace@1175..1176 " "
                                        PATTERN_VARIABLE@1176..1178
                                          Lident@1176..1177 "a"
                                          Whitespace@1177..1178 " "
                                        Eq@1178..1179 "="
                                        Whitespace@1179..1180 " "
                                        EXPR_LET_VALUE@1180..1201
                                          EXPR_CALL@1180..1201
                                            EXPR_LIDENT@1180..1184
                                              Lident@1180..1184 "flip"
                                            ARG_LIST@1184..1201
                                              LParen@1184..1185 "("
                                              ARG@1185..1199
                                                EXPR_CALL@1185..1199
                                                  EXPR_LIDENT@1185..1192
                                                    Lident@1185..1192 "x_add_1"
                                                  ARG_LIST@1192..1199
                                                    LParen@1192..1193 "("
                                                    ARG@1193..1198
                                                      EXPR_LIDENT@1193..1198
                                                        Lident@1193..1198 "start"
                                                    RParen@1198..1199 ")"
                                              RParen@1199..1200 ")"
                                              Whitespace@1200..1201 " "
                                        InKeyword@1201..1203 "in"
                                        Whitespace@1203..1208 "\n    "
                                        EXPR_LET_BODY@1208..1262
                                          EXPR_LET@1208..1262
                                            LetKeyword@1208..1211 "let"
                                            Whitespace@1211..1212 " "
                                            PATTERN_WILDCARD@1212..1214
                                              WildcardKeyword@1212..1213 "_"
                                              Whitespace@1213..1214 " "
                                            Eq@1214..1215 "="
                                            Whitespace@1215..1216 " "
                                            EXPR_LET_VALUE@1216..1251
                                              EXPR_CALL@1216..1251
                                                EXPR_LIDENT@1216..1230
                                                  Lident@1216..1230 "string_println"
                                                ARG_LIST@1230..1251
                                                  LParen@1230..1231 "("
                                                  ARG@1231..1249
                                                    EXPR_CALL@1231..1249
                                                      EXPR_LIDENT@1231..1246
                                                        Lident@1231..1246 "point_to_string"
                                                      ARG_LIST@1246..1249
                                                        LParen@1246..1247 "("
                                                        ARG@1247..1248
                                                          EXPR_LIDENT@1247..1248
                                                            Lident@1247..1248 "a"
                                                        RParen@1248..1249 ")"
                                                  RParen@1249..1250 ")"
                                                  Whitespace@1250..1251 " "
                                            InKeyword@1251..1253 "in"
                                            Whitespace@1253..1259 "\n\n    "
                                            EXPR_LET_BODY@1259..1262
                                              EXPR_UNIT@1259..1262
                                                LParen@1259..1260 "("
                                                RParen@1260..1261 ")"
                                                Whitespace@1261..1262 "\n"
      RBrace@1262..1263 "}"
      Whitespace@1263..1264 "\n"