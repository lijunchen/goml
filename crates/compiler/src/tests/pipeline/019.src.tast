fn make_point() -> Point {
  Point { x: 0, y: 0 }
}

fn flip(point/0: Point) -> Point {
  let Point { x: x/1: int, y: y/2: int } = (point/0 : Point) in
  Point { x: (y/2 : int), y: (x/1 : int) }
}

fn wrap_int(x/3: int) -> Wrapper[int] {
  Wrapper { value: (x/3 : int) }
}

fn x_add_1(p/4: Point) -> Point {
  let Point { x: x/5: int, y: y/6: int } = (p/4 : Point) in
  Point { x: (x/5 : int) + 1, y: (y/6 : int) }
}

fn get_x(p/7: Point) -> int {
  let Point { x: x/8: int, y: y/9: int } = (p/7 : Point) in
  (x/8 : int)
}

fn get_y(p/10: Point) -> int {
  let Point { x: x/11: int, y: y/12: int } = (p/10 : Point) in
  (y/12 : int)
}

fn point_to_string(p/13: Point) -> string {
  let Point { x: x/14: int, y: y/15: int } = (p/13 : Point) in
  (string_add : (string, string) -> string)((string_add : (string, string) -> string)((string_add : (string, string) -> string)("Point { x: ", (int_to_string : (int) -> string)((x/14 : int))), (string_add : (string, string) -> string)(", y: ", (int_to_string : (int) -> string)((y/15 : int)))), "}")
}

fn main() -> unit {
  let start/16: Point = (make_point : () -> Point)() in
  let _ : unit = (string_println : (string) -> unit)((point_to_string : (Point) -> string)((start/16 : Point))) in
  let swapped/17: Point = (flip : (Point) -> Point)(Point { x: 1, y: 2 }) in
  let _ : unit = (string_println : (string) -> unit)((point_to_string : (Point) -> string)((swapped/17 : Point))) in
  let boxed/18: Wrapper[int] = (wrap_int : (int) -> Wrapper[int])(3) in
  let point_box/19: Wrapper[Point] = Wrapper { value: (swapped/17 : Point) } in
  let a/20: Point = (x_add_1 : (Point) -> Point)((start/16 : Point)) in
  let _ : unit = (string_println : (string) -> unit)((point_to_string : (Point) -> string)((a/20 : Point))) in
  let a/21: Point = (flip : (Point) -> Point)((x_add_1 : (Point) -> Point)((start/16 : Point))) in
  let _ : unit = (string_println : (string) -> unit)((point_to_string : (Point) -> string)((a/21 : Point))) in
  ()
}