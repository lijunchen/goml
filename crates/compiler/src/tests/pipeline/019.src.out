Point { x: 0, y: 0}
Point { x: 2, y: 1}
Point { x: 1, y: 0}
Point { x: 0, y: 1}
